src/EditorFeatures/Core/Tagging/TaggerThreadCoordinator.cs

```diff
@@ -0,0 +1,192 @@
+            /// </summary>
+            public void PerformWork()
+            {
+                if (_cancellationToken.IsCancellationRequested)
+                {
+                    // Reasonable chance the tagger was canceled between when it enqueued work and when we're running.
+                    // Just clear it out and return.  Fast path that to avoid any actual async/await task work.
+                    _completionSource.TrySetCanceled(_cancellationToken);
+                    _registration.Dispose();
+                }
+                else
+                {
+                    // Otherwise, kick off the work and have is signal to the originator (through the completion source)
+                    // when it is finished.  We do this in a fire-and-forget fashion.  The async tagging work the queue
+                    // is doing shouldn't impact the ability for the threading coordinator to move to the next batch of
+                    // work (once the tagging work moves to the bg).
+                    _ = PerformWorkAsync();
+                }
+            }
+
+            private async Task PerformWorkAsync()
+            {
+                try
+                {
+                    // Go and actually do the work, forwarding the progress to the completion source to let the tagger know when finised.
+                    await _work(_cancellationToken).ConfigureAwait(true);
+                    _completionSource.TrySetResult(true);
+                }
+                catch (OperationCanceledException ex)
+                {
+                    _completionSource.TrySetCanceled(ex.CancellationToken);
+                }
+                catch (Exception ex)
+                {
+                    _completionSource.TrySetException(ex);
+                }
+                finally
+                {
+                    // Final cleanup so we don't leak registrations.
+                    _registration.Dispose();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Context we use to get back to the UI thread, and to know if the host is shutting down.
+        /// </summary>
+        private readonly IThreadingContext _threadingContext;
+
+        /// <summary>
+        /// Queue that batches up all the work we've been requested to do so we can execute all of it at once in one
+        /// single UI thread message.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<TaggerWork> _queue;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public TaggerThreadCoordinator(
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListenerProvider listenerProvider)
+        {
+            _threadingContext = threadingContext;
+
+            _queue = new AsyncBatchingWorkQueue<TaggerWork>(
+                TaggerDelay.NearImmediate.ComputeTimeDelay(),
```

Comment:
📝 This should be zero, similar to `FileChangeWatcher`

```suggestion
                TimeSpan.Zero,
```

Reply:
interesting...  i'm slightly concerned by that.  Is there a downside to using near-immediate to at least collect some reasonably large chunk of items?

Reply:
Yes, we already have the delays built into the tagger. Including a delay here would serve no purpose other than to increase the delays we already set and confuse the system for determining when things happen.

Reply:
@CyrusNajmabadi `TimeSpan.Zero` already provides batching benefits as shown by [link coming]. This issue needs to be addressed.